### PR TITLE
Speed up rotations by caching Pauli matrices

### PR DESCRIPTION
**Pull request by xbe**
Before:
```bash
$ python3 -m timeit -s "from src.lib import ops; import math" "ops.RotationX(math.pi/2)"
5000 loops, best of 5: 44.2 usec per loop
```

After:
```bash
$ python3 -m timeit -s "from src.lib import ops; import math" "ops.RotationX(math.pi/2)"
10000 loops, best of 5: 37.3 usec per loop
```

That's a nice ~15% boost!

### CODE REVIEW

**Comment by xbe** [comment]
PTAL, @qcc4cp.

File: src/lib/ops.py [hunk 1]
```diff
@@ -184,6 +184,9 @@ def U1(lam):
                             (0.0, cmath.exp(1j * lam))]))
 
 
+_PAULI_VECTOR = np.array([PauliX(), PauliY(), PauliZ()])
```
Comment (xbe) [inline src/lib/ops.py:187]: Perhaps this should be named `_PAULI_TENSOR` to be more accurate, WDYT?

Comment (xbe) [reply src/lib/ops.py]: I renamed it.

**Comment by qcc4cp** [comment]
See comment on the performance / teaching trade-off, and where I suspect the speedup comes from. Thanks.

File: src/lib/ops.py [hunk 2]
```diff
@@ -196,7 +199,7 @@ def Rotation(v, theta):
     raise ValueError('Rotation vector v must be a 3D real unit vector.')
 
   return np.cos(theta / 2) * Identity() - 1j * np.sin(theta / 2) * (
-      v[0] * PauliX() + v[1] * PauliY() + v[2] * PauliZ())
+      np.tensordot(v, _PAULI_VECTOR, axes=1))
```
[the remark this answers is not in the document]
Comment (xbe) [reply src/lib/ops.py]: Great intuition! Explicit multiplication with cached Paulis ends up being even faster:
```bash
$ python3 -m timeit -s "from src.lib import ops; import math" "ops.RotationX(math.pi/2)"
10000 loops, best of 5: 30.5 usec per loop
```

I think there is a strong case for caching matrices as "singletons" and making them available (not marking for internal use with trailing underscore):
1. Speed improvement.
2. Memory improvement.
3. Not having to call the functions leads to cleaner syntax. Compare:
```python
# with functions
PauliX()(PauliY())

# with "singletons"
PAULI_X(PAULI_Y)
```
4. Cirq does it.

One con to the singleton approach that I can think of is that someone can accidentally modify the matrices. I don't think it's likely enough to cause concern, but there's always `PAULI_X.flags.writeable = False`.

File: src/lib/ops.py [hunk 3]
```diff
@@ -195,12 +201,12 @@ def Rotation(v: np.ndarray, theta: float) -> np.ndarray:
   """Produce the single-qubit rotation operator."""
 
   v = np.asarray(v)
-  if (v.shape != (3,) or not np.isclose(v @ v, 1) or
+  if (v.shape != (3,) or not math.isclose(v @ v, 1) or
```
Comment (xbe) [inline src/lib/ops.py:204]: Lesson learned: `math.isclose()` is 300 times faster than `np.isclose()` for scalars:
```
$ python3 -m timeit -s "import numpy as np; import math" "np.isclose(1.0001, 1)"                                                                                                                                                                       
20000 loops, best of 5: 15.4 usec per loop
$ python3 -m timeit -s "import numpy as np; import math" "math.isclose(1.0001, 1)"                                                                                                                                                                     
5000000 loops, best of 5: 48.4 nsec per loop
```

See: https://github.com/numpy/numpy/issues/16160 .

**Comment by qcc4cp** [comment]
I could merge now, but I think it would be nice if you added a few of your observations as comments into the code. Perhaps something like:
   # xbe@: Benchmarking shows that for scalars math.isclose is 300x faster than np.isclose.

and
  # xbe@: We cache the Pauli matrices here for performance reasons. Also, the 
  #            syntax is nicer ... (your content)

You know what I mean. I can't add you as a co-author to the book, but I'd want to give you as much visibility as possible. You could also use your real name for this purpose, instead of xbe.

**Pull request merged**